core-bundle/src/EventListener/DataContainer/RootPageDependentSelectListener.php

```diff
@@ -0,0 +1,207 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Contao\CoreBundle\ServiceAnnotation\Callback;
+use Contao\CoreBundle\ServiceAnnotation\Hook;
+use Contao\DataContainer;
+use Contao\Image;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+class RootPageDependentSelectListener
+{
+    private Connection $connection;
+    private TranslatorInterface $translator;
+    private ScopeMatcher $scopeMatcher;
+    private RequestStack $requestStack;
+    private CsrfTokenManagerInterface $csrfTokenManager;
+    private string $csrfTokenName;
+
+    public function __construct(Connection $connection, TranslatorInterface $translator, ScopeMatcher $scopeMatcher, RequestStack $requestStack, CsrfTokenManagerInterface $csrfTokenManager, string $csrfTokenName)
+    {
+        $this->connection = $connection;
+        $this->translator = $translator;
+        $this->scopeMatcher = $scopeMatcher;
+        $this->requestStack = $requestStack;
+        $this->csrfTokenManager = $csrfTokenManager;
+        $this->csrfTokenName = $csrfTokenName;
+    }
+
+    /**
+     * @Hook("loadDataContainer")
+     */
+    public function onLoadDataContainer(string $table): void
```

Comment:
Why is the logic not in the widget itself? It's a very specific widget that loads stuff per root page. Why setting them from the outside if the widget itself already knows about them?

Reply:
Valid point. This has possibly historical reasons how this thing evolved 😄

Reply:
I've restructured it in c742a6b6e.